sdk/go/pulumix/compose.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumix
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"runtime/debug"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/internal"
+)
+
+// Composer allows composing multiple outputs together in a type-safe manner.
+//
+// To use Composer, begin a composition with the [Compose] function.
+// This returns an Output[O] that will be fulfilled with the value returned by f.
+//
+//	o := pulumix.Compose(ctx, func(c *pulumix.Composer) (O, error) {
+//		// ...
+//	})
+//
+// Inside the function, use [ComposeAwait] to await other outputs.
+//
+//	v1 := pulumix.ComposeAwait(c, o1)
+//	v2 := pulumix.ComposeAwait(c, o2)
+//	// ...
+//
+// Combine the values into a single result and return it.
+// For example:
+//
+//	return &MyStruct{
+//		Foo: v1,
+//		Bar: v2,
+//		...
+//	}, nil
+//
+// If f returns an error, the output will be rejected.
+// If any of the awaited outputs fail, the composition will abort
+// and the output will be rejected.
+//
+// Composer allows combining an arbitrary number of outputs.
+// For simpler cases, consider [Apply] or any of the ApplyN variants.
+type Composer struct {
+	ctx         context.Context
+	known       bool
+	secret      bool
+	deps        []internal.Resource
+	outputState *internal.OutputState
+
+	fulfilled bool // true if the output has been fulfilled
+}
+
+// Compose begins a new output composition operation.
+//
+//	o := pulumix.Compose[O](ctx, func(c *pulumix.Composer) (O, error) {
+//		// ...
+//	})
+//
+// Inside f, use [ComposeAwait] to await other outputs.
+// The value returned by f is the value of the output.
+// If f returns an error, the output will be rejected.
+//
+// Compose has some restrictions:
+//
+//   - f MUST call ComposeAwait in the same goroutine.
+//   - f MUST NOT use the Composer after it returns.
+//   - f SHOULD NOT spawn new goroutines.
+//     If it does, it MUST NOT use the Composer in those goroutines.
+func Compose[T any](ctx context.Context, f func(*Composer) (T, error)) Output[T] {
+	var wg internal.WorkGroup
```

Comment:
Is it ok to have brand new WorkGroup instance for this output?

Should we encourage an API that takes the WorkGroup from a pulumi.Context? 

https://github.com/pulumi/pulumi/blob/67f127da9e8c6392c85391310970b97ba9732bf5/sdk/go/pulumi/context.go#L53-#L53

Reply:
There's a comment here https://github.com/pulumi/pulumi/blob/67f127da9e8c6392c85391310970b97ba9732bf5/sdk/go/internal/types.go#L194-#L194

Reply:
The observable effect of this, is that if you do fmt.Println() inside `f func(*Composer) (T, error)`, the program should wait until that effect executes even if the output takes awhile to resolve and nothing else waits on it. Also if there's an error inside the Composer the program should catch the error. I think the waitGroup machinery needs to link this new output to the pulumi.Context for these behaviors to happen.

Reply:
The observable effect of this, is that if you do fmt.Println() inside `f func(*Composer) (T, error)`, the program should wait until that effect executes even if the output takes awhile to resolve and nothing else waits on it. Also if there's an error inside the Composer the program should catch the error. I think the waitGroup machinery needs to link this new output to the pulumi.Context for these behaviors to happen.

Reply:
I'm not certain if Compose should depend on `pulumi.Context`, but I don't feel strongly about it.
Right now, the side-effects inside the composition function will come into play only when you await the composition function.

Reply:
Yes, that'd work I think!

Reply:
@t0yv0 Pushed a commit to do that; it tracks the pulumi.Context's work group on the context.Context returned by `ctx.Context()`. If the work group is available, `Compose` will use that, effectively blocking `Run` until the side effects inside the composition function have had a chance to run.